Build docker incrementally by default

vdemeester opened the pull request:
This speeds up docker build time drastically. It still possible to disable this by setting `DOCKER_INCREMENTAL_BUILD` to `0` (and this is what should be done on the CI).

/cc @icecrime @tianon @cpuguy83 @aaronlehmann 

This PR doesn't change anything for the CI, i.e. the CI still builds it using `-a`, a.k.a `force rebuilding of packages that are already up-to-date.`.

🐸

Signed-off-by: Vincent Demeester <vincent@sbr.pm>

vdemeester commented on hack/make.sh:
```diff
@@ -159,7 +159,7 @@ ORIG_BUILDFLAGS=( -tags "autogen netgo static_build sqlite_omit_load_extension $
 # builds by installing dependent packages to the GOPATH.
 REBUILD_FLAG="-a"
-if [ "$DOCKER_INCREMENTAL_BINARY" ]; then
+if [ "$DOCKER_INCREMENTAL_BINARY" == "1" ] || [ "$DOCKER_INCREMENTAL_BINARY" == "true" ]; then
```
Not sure I should check for anything other than `1` but :sweat_smile:

AkihiroSuda replied:
I think we should limit it to 1, unless other options supports true as well

cpuguy83 commented on Makefile:
```diff
@@ -21,7 +22,7 @@ DOCKER_ENVS := \
 	-e DOCKER_GITCOMMIT \
 	-e DOCKER_GRAPHDRIVER=$(DOCKER_GRAPHDRIVER) \
-	-e DOCKER_INCREMENTAL_BINARY \
+	-e DOCKER_INCREMENTAL_BINARY=$(DOCKER_INCREMENTAL_BINARY) \
```
How come this change is needed?
Should be equivalent.

thaJeztah replied:
Looks like we can remove it for `DOCKER_GRAPHDRIVER` as well then

vdemeester replied:
actually it's not equivalent because we *set* `DOCKER_INCREMENTAL_BINARY` (and same goes for `DOCKER_GRAPHDRIVER`) in the `Makefile` – and thus the `docker run` command doesn't pick it up :sweat:

cpuguy83 replied:
Because it's not exported? Makes sense.

vdemeester replied:
@tianon because I didn't really though of that :stuck_out_tongue_closed_eyes:

vdemeester replied:
and Done :)

cpuguy83 commented:
1 nit, but LGTM.

vdemeester committed:
Build docker incrementally by default

This speeds up docker build time drastically. It still possible to
disable this by setting `DOCKER_INCREMENTAL_BUILD` to `0` (and this is
what should be done on the CI).

Signed-off-by: Vincent Demeester <vincent@sbr.pm>

tianon reviewed (approved):
:+1:

LK4D4 commented:
btw windows has unrelated failure:
```
16:32:47 ----------------------------------------------------------------------
16:32:47 FAIL: docker_cli_ps_test.go:309: DockerSuite.TestPsListContainersFilterAncestorImage
16:32:47 
16:32:47 docker_cli_ps_test.go:380:
16:32:47     checkPsAncestorFilterOutput(c, out, filter.filterName, filter.expectedIDs)
16:32:47 docker_cli_ps_test.go:396:
16:32:47     c.Assert(actualIDs, checker.HasLen, len(expectedIDs), check.Commentf("Expected filtered container(s) for %s ancestor filter to be %v:%v, got %v:%v", filterName, len(expectedIDs), expectedIDs, len(actualIDs), actualIDs))
16:32:47 ... obtained []string = []string{}
16:32:47 ... n int = 1
16:32:47 ... Expected filtered container(s) for sha256:fe9ff0cb9b9dd5b636bbd0222c5d678abefc206d94e11da239dc1f2fb92bdf00
16:32:47 runtime: failed to create new OS thread (have ancestor filter to be 1:[d22a2c34a537175da06b3dddea65a85da4f27667edfe947390c59088a099a505], got 0:[]
```
ping @jhowardmsft

LK4D4 commented on Makefile:
```diff
@@ -2,6 +2,9 @@
 # set the graph driver as the current graphdriver if not set
 DOCKER_GRAPHDRIVER := $(if $(DOCKER_GRAPHDRIVER),$(DOCKER_GRAPHDRIVER),$(shell docker info 2>&1 | grep "Storage Driver" | sed 's/.*: //'))
+export DOCKER_GRAPHDRIVER
+DOCKER_INCREMENTAL_BINARY := $(if $(DOCKER_INCREMENTAL_BINARY),$(DOCKER_INCREMENTAL_BINARY),1)
```
what does this mean? :)

vdemeester replied:
It set's `DOCKER_INCREMENTAL_BINARY` to `1` if not already specified as a enviroment variable.
This allows you to do `make DOCKER_INCREMENTAL_BINARY=0 binary` to disable it for example.

LK4D4 commented:
LGTM

vdemeester commented:
:tada:
